File: java/leetcode/src/main/java/com/leetcode/tree/Path_Sum_II_113.java
package com.leetcode.tree;

import java.util.ArrayList;
import java.util.List;

public class Path_Sum_II_113 {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer> path = new ArrayList<>();
        dfs(root, targetSum, path);
        return res;

    }

    void dfs(TreeNode node, int left, List<Integer> path) {
        if (node == null) return;
        path.add(node.val);
        if (isLeaf(node) && node.val == left) {
            res.add(new ArrayList<Integer>(path));
        } else {
            dfs(node.left, left - node.val, path);
            dfs(node.right, left - node.val, path);
        }
        path.remove(path.size() - 1);

    }

    boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }
}

File: js/algrithm/tree/Path_Sum_II_113.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {
    const ans = []
    const dfs = function(node, acc, path) {
     if (!node) return
     path.push(node.val)
     if (!node.left && !node.right) {
         if (acc + node.val === targetSum) ans.push(path.slice())
     }
     dfs(node.left, acc + node.val, path)
     dfs(node.right, acc + node.val, path)
     path.pop()
    }
 
    dfs(root, 0, [])
    return ans
 };